### PR TITLE
[hail/ptypes] infer Coalesce PType

### PR DESCRIPTION
**Pull request by akotlar**
Partially addresses #6952

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/InferPType.scala [hunk 1]
```diff
@@ -414,6 +416,21 @@ object InferPType {
 
         branchType.setRequired(branchType.required && cond.pType2.required)
       }
+
+      case Coalesce(values) => {
+        InferPType(values(0), env)
+        var pTypes = ArrayBuffer[PType](values(0)._pType2)
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/InferPType.scala]: I think this is clearer if written functionally:

```scala
 getNestedElementPTypes(ptypes = values.map { v => 
    InferPType(v, env)
    v.pType2
})
```

Comment (akotlar) [reply hail/src/main/scala/is/hail/expr/ir/InferPType.scala]: What about the assertion that all are isOfType?

Comment (akotlar) [reply hail/src/main/scala/is/hail/expr/ir/InferPType.scala]: I could write this with a branch in the map, e.g

```scala
var head: IR: _
var i = 0;
 getNestedElementPTypes(ptypes = values.map { v => 
    InferPType(v, env)
    if(i > 0) assert(v._pType2 isOfType head._pType2)
    v.pType2
    i += 1
})
```

But I think this is less clear, this mixture of imperative and functional code.

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/InferPType.scala]: getNestedElementPTypes should do that assertion, and it should probably be an assertion that all the virtual types are the same.

**Comment by akotlar** [comment]
> oops, forgot to request changes.
> 
> assertion should move into getNestedElementPTypes as something like the following:
> 
> ```scala
> val vTypes = values.iterator.map(_.virtualType).toSet
> if (vTypes.size != 1)
>     throw new AssertionError(s"invalid call to getNestedElementPTypes: $vTypes")
> ```

Should we check isOfType instead of virtual type, for case when we say try to coalesce canonical and non-canonical implementations.

**Comment by tpoterba** [comment]
I'm not really sure what `isOfType` is supposed to do -- it seems like it's meant to check that the virtual types are the same, so we should just do that (and remove isOfType)`

**Comment by akotlar** [comment]
It checks that the physical types are the same, excepting requiredeness. For instance that two PStruct are instances of the same class (PStruct), and have the same field names, indices, and those fields are isOfType the same PType

**Comment by tpoterba** [comment]
> It checks that the physical types are the same, excepting requiredeness. For instance that two PStruct are instances of the same class (PStruct), and have the same field names, indices, and those fields are isOfType the same PType

I don't think that's the behavior we want in this check -- you can have a `Coalesce(x, y)` where `x` is of type PWeirdStruct and `y` is of type PCanonicalStruct. That's totally fine.

**Comment by tpoterba** [comment]
after looking at the usages, I'm now quite convinced that this should be renamed to `sameVirtualType`, with the implementation correspondingly changed.

There are a few usages that are wrong, and should instead be ptype equality.

File: hail/src/main/scala/is/hail/expr/ir/InferPType.scala [hunk 2]
```diff
@@ -4,35 +4,42 @@ import is.hail.expr.types.physical._
 import is.hail.expr.types.virtual.{TNDArray, TTuple}
 import is.hail.utils._
 
+import scala.collection.mutable.ArrayBuffer
+
 object InferPType {
   def getNestedElementPTypes(ptypes: Seq[PType]): PType = {
+    assert(ptypes.forall(_.virtualType.isOfType(ptypes.head.virtualType)))
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/ir/InferPType.scala]: top level since isOfType is a recursive check.

**Pull request merged**